server/ws/websockets.go

```diff
@@ -154,20 +109,123 @@ func (ws *Server) handleWebSocketOnChange(w http.ResponseWriter, r *http.Request
 		}
 
 		switch command.Action {
+		case "AUTH":
+			log.Printf(`Command: AUTH, client: %s`, client.RemoteAddr())
+			ws.authenticateListener(&wsSession, command.Token)
+
 		case "ADD":
 			log.Printf(`Command: Add blockID: %v, client: %s`, command.BlockIDs, client.RemoteAddr())
-			ws.AddListener(client, command.BlockIDs)
+			ws.addListener(&wsSession, command.BlockIDs)
 
 		case "REMOVE":
 			log.Printf(`Command: Remove blockID: %v, client: %s`, command.BlockIDs, client.RemoteAddr())
-			ws.RemoveListenerFromBlocks(client, command.BlockIDs)
+			ws.removeListenerFromBlocks(&wsSession, command.BlockIDs)
 
 		default:
 			log.Printf(`ERROR webSocket command, invalid action: %v`, command.Action)
 		}
 	}
 }
 
+func (ws *Server) isValidSessionToken(token string) bool {
+	if ws.singleUser {
+		return true
+	}
+
+	session, err := ws.auth.GetSession(token)
+	if session == nil || err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (ws *Server) authenticateListener(wsSession *websocketSession, token string) {
+	isValidSession := ws.isValidSessionToken(token)
+	if !isValidSession {
+		wsSession.client.Close()
+		return
+	}
+
+	// Authenticated
+	wsSession.isAuthenticated = true
+	log.Printf("authenticateListener: Authenticated")
+}
+
+// AddListener adds a listener for a block's change.
+func (ws *Server) addListener(wsSession *websocketSession, blockIDs []string) {
+	if !wsSession.isAuthenticated {
+		log.Printf("addListener: NOT AUTHENTICATED")
```

Comment:
Maybe sending a response like `{"error": "NOT AUTHENTICATED"}`?

Reply:
Good call. Added an error message to the client.